{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 31,
   "metadata": {},
   "outputs": [],
   "source": [
    "import boto3\n",
    "import pysam\n",
    "import os\n",
    "import pandas as pd"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Setup"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 51,
   "metadata": {},
   "outputs": [],
   "source": [
    "os.chdir('/data/jake/sv-gmm/python')\n",
    "sv_tbl='../data/sv-modes.tsv'\n",
    "dirout='../data/variants'\n",
    "idx_1kg_phase3='../data/20130502.phase3.low_coverage.alignment.index' # from: aws s3 ls s3://1000genomes/alignment_indices/\n",
    "if not os.path.exists(dirout):\n",
    "    os.mkdir(dirout)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 34,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>SV ID</th>\n",
       "      <th># modes predicted</th>\n",
       "      <th>chr</th>\n",
       "      <th>start</th>\n",
       "      <th>stop</th>\n",
       "      <th>allele frequency</th>\n",
       "      <th>length</th>\n",
       "      <th>Total # samples</th>\n",
       "      <th># reference samples</th>\n",
       "      <th>Mode 1</th>\n",
       "      <th>Mode 2</th>\n",
       "      <th>Mode 3</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>UW_VH_9038</td>\n",
       "      <td>1</td>\n",
       "      <td>19</td>\n",
       "      <td>54887338</td>\n",
       "      <td>54888354</td>\n",
       "      <td>0.014</td>\n",
       "      <td>1016</td>\n",
       "      <td>51</td>\n",
       "      <td>0</td>\n",
       "      <td>['HG01125', 'HG01896', 'HG01956', 'HG02009', '...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>SI_BD_10797</td>\n",
       "      <td>1</td>\n",
       "      <td>10</td>\n",
       "      <td>86801825</td>\n",
       "      <td>86802449</td>\n",
       "      <td>0.033</td>\n",
       "      <td>624</td>\n",
       "      <td>283</td>\n",
       "      <td>139</td>\n",
       "      <td>['HG00154', 'HG00159', 'HG00250', 'HG00264', '...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>UW_VH_19141</td>\n",
       "      <td>1</td>\n",
       "      <td>3</td>\n",
       "      <td>177294474</td>\n",
       "      <td>177297489</td>\n",
       "      <td>0.032</td>\n",
       "      <td>3015</td>\n",
       "      <td>133</td>\n",
       "      <td>1</td>\n",
       "      <td>['HG00180', 'HG00185', 'HG00189', 'HG00237', '...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>DEL_pindel_47187</td>\n",
       "      <td>2</td>\n",
       "      <td>18</td>\n",
       "      <td>45379612</td>\n",
       "      <td>45379612</td>\n",
       "      <td>0.550</td>\n",
       "      <td>195</td>\n",
       "      <td>847</td>\n",
       "      <td>66</td>\n",
       "      <td>['HG00114', 'HG00132', 'HG00142', 'HG00150', '...</td>\n",
       "      <td>['HG00108', 'HG00111', 'HG00121', 'HG00125', '...</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>DEL_pindel_24042</td>\n",
       "      <td>2</td>\n",
       "      <td>7</td>\n",
       "      <td>136996507</td>\n",
       "      <td>136996739</td>\n",
       "      <td>0.071</td>\n",
       "      <td>232</td>\n",
       "      <td>184</td>\n",
       "      <td>48</td>\n",
       "      <td>['HG01241', 'HG01259', 'HG01488', 'HG01890', '...</td>\n",
       "      <td>['HG01125', 'HG01392', 'HG01403', 'HG01556', '...</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "              SV ID  # modes predicted  chr      start       stop  \\\n",
       "0        UW_VH_9038                  1   19   54887338   54888354   \n",
       "1       SI_BD_10797                  1   10   86801825   86802449   \n",
       "2       UW_VH_19141                  1    3  177294474  177297489   \n",
       "3  DEL_pindel_47187                  2   18   45379612   45379612   \n",
       "4  DEL_pindel_24042                  2    7  136996507  136996739   \n",
       "\n",
       "   allele frequency  length  Total # samples  # reference samples  \\\n",
       "0             0.014    1016               51                    0   \n",
       "1             0.033     624              283                  139   \n",
       "2             0.032    3015              133                    1   \n",
       "3             0.550     195              847                   66   \n",
       "4             0.071     232              184                   48   \n",
       "\n",
       "                                              Mode 1  \\\n",
       "0  ['HG01125', 'HG01896', 'HG01956', 'HG02009', '...   \n",
       "1  ['HG00154', 'HG00159', 'HG00250', 'HG00264', '...   \n",
       "2  ['HG00180', 'HG00185', 'HG00189', 'HG00237', '...   \n",
       "3  ['HG00114', 'HG00132', 'HG00142', 'HG00150', '...   \n",
       "4  ['HG01241', 'HG01259', 'HG01488', 'HG01890', '...   \n",
       "\n",
       "                                              Mode 2 Mode 3  \n",
       "0                                                NaN    NaN  \n",
       "1                                                NaN    NaN  \n",
       "2                                                NaN    NaN  \n",
       "3  ['HG00108', 'HG00111', 'HG00121', 'HG00125', '...    NaN  \n",
       "4  ['HG01125', 'HG01392', 'HG01403', 'HG01556', '...    NaN  "
      ]
     },
     "execution_count": 34,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df = pd.read_csv(sv_tbl,sep='\\t')\n",
    "df.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 35,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>id</th>\n",
       "      <th>n_modes</th>\n",
       "      <th>chr</th>\n",
       "      <th>start</th>\n",
       "      <th>stop</th>\n",
       "      <th>allele_freq</th>\n",
       "      <th>length</th>\n",
       "      <th>n_samples</th>\n",
       "      <th>n_ref_samples</th>\n",
       "      <th>mode_1</th>\n",
       "      <th>mode_2</th>\n",
       "      <th>mode_3</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>UW_VH_9038</td>\n",
       "      <td>1</td>\n",
       "      <td>19</td>\n",
       "      <td>54887338</td>\n",
       "      <td>54888354</td>\n",
       "      <td>0.014</td>\n",
       "      <td>1016</td>\n",
       "      <td>51</td>\n",
       "      <td>0</td>\n",
       "      <td>['HG01125', 'HG01896', 'HG01956', 'HG02009', '...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>SI_BD_10797</td>\n",
       "      <td>1</td>\n",
       "      <td>10</td>\n",
       "      <td>86801825</td>\n",
       "      <td>86802449</td>\n",
       "      <td>0.033</td>\n",
       "      <td>624</td>\n",
       "      <td>283</td>\n",
       "      <td>139</td>\n",
       "      <td>['HG00154', 'HG00159', 'HG00250', 'HG00264', '...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>UW_VH_19141</td>\n",
       "      <td>1</td>\n",
       "      <td>3</td>\n",
       "      <td>177294474</td>\n",
       "      <td>177297489</td>\n",
       "      <td>0.032</td>\n",
       "      <td>3015</td>\n",
       "      <td>133</td>\n",
       "      <td>1</td>\n",
       "      <td>['HG00180', 'HG00185', 'HG00189', 'HG00237', '...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>DEL_pindel_47187</td>\n",
       "      <td>2</td>\n",
       "      <td>18</td>\n",
       "      <td>45379612</td>\n",
       "      <td>45379612</td>\n",
       "      <td>0.550</td>\n",
       "      <td>195</td>\n",
       "      <td>847</td>\n",
       "      <td>66</td>\n",
       "      <td>['HG00114', 'HG00132', 'HG00142', 'HG00150', '...</td>\n",
       "      <td>['HG00108', 'HG00111', 'HG00121', 'HG00125', '...</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>DEL_pindel_24042</td>\n",
       "      <td>2</td>\n",
       "      <td>7</td>\n",
       "      <td>136996507</td>\n",
       "      <td>136996739</td>\n",
       "      <td>0.071</td>\n",
       "      <td>232</td>\n",
       "      <td>184</td>\n",
       "      <td>48</td>\n",
       "      <td>['HG01241', 'HG01259', 'HG01488', 'HG01890', '...</td>\n",
       "      <td>['HG01125', 'HG01392', 'HG01403', 'HG01556', '...</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                 id  n_modes  chr      start       stop  allele_freq  length  \\\n",
       "0        UW_VH_9038        1   19   54887338   54888354        0.014    1016   \n",
       "1       SI_BD_10797        1   10   86801825   86802449        0.033     624   \n",
       "2       UW_VH_19141        1    3  177294474  177297489        0.032    3015   \n",
       "3  DEL_pindel_47187        2   18   45379612   45379612        0.550     195   \n",
       "4  DEL_pindel_24042        2    7  136996507  136996739        0.071     232   \n",
       "\n",
       "   n_samples  n_ref_samples  \\\n",
       "0         51              0   \n",
       "1        283            139   \n",
       "2        133              1   \n",
       "3        847             66   \n",
       "4        184             48   \n",
       "\n",
       "                                              mode_1  \\\n",
       "0  ['HG01125', 'HG01896', 'HG01956', 'HG02009', '...   \n",
       "1  ['HG00154', 'HG00159', 'HG00250', 'HG00264', '...   \n",
       "2  ['HG00180', 'HG00185', 'HG00189', 'HG00237', '...   \n",
       "3  ['HG00114', 'HG00132', 'HG00142', 'HG00150', '...   \n",
       "4  ['HG01241', 'HG01259', 'HG01488', 'HG01890', '...   \n",
       "\n",
       "                                              mode_2 mode_3  \n",
       "0                                                NaN    NaN  \n",
       "1                                                NaN    NaN  \n",
       "2                                                NaN    NaN  \n",
       "3  ['HG00108', 'HG00111', 'HG00121', 'HG00125', '...    NaN  \n",
       "4  ['HG01125', 'HG01392', 'HG01403', 'HG01556', '...    NaN  "
      ]
     },
     "execution_count": 35,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df.columns = ['id', 'n_modes', 'chr', 'start', 'stop', 'allele_freq', 'length', 'n_samples', 'n_ref_samples', 'mode_1', 'mode_2', 'mode_3']\n",
    "df.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 36,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "['UW_VH_9038',\n",
       " 'SI_BD_10797',\n",
       " 'UW_VH_19141',\n",
       " 'DEL_pindel_47187',\n",
       " 'DEL_pindel_24042',\n",
       " 'BI_GS_DEL1_B2_P0106_507',\n",
       " 'BI_GS_DEL1_B4_P2674_173',\n",
       " 'UW_VH_10394',\n",
       " 'BI_GS_DEL1_B2_P0114_484']"
      ]
     },
     "execution_count": 36,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "variants=df.id.tolist()\n",
    "variants"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 43,
   "metadata": {},
   "outputs": [],
   "source": [
    "# output dirs \n",
    "# variant_name/\n",
    "# |\n",
    "# l_ samplot/\n",
    "# l_ bam/\n",
    "for v in variants:\n",
    "    d = os.path.join(dirout,v)\n",
    "    if not os.path.exists(d):\n",
    "        os.mkdir(d)\n",
    "    d_bam = os.path.join(dirout,v,'bam')\n",
    "    d_samplot = os.path.join(dirout,v,'samplot')\n",
    "    if not os.path.exists(d_bam):\n",
    "        os.mkdir(d_bam)\n",
    "    if not os.path.exists(d_samplot):\n",
    "        os.mkdir(d_samplot)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 81,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "(2535,)\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "0    data/HG00096/alignment/HG00096.mapped.ILLUMINA...\n",
       "1    data/HG00097/alignment/HG00097.mapped.ILLUMINA...\n",
       "2    data/HG00099/alignment/HG00099.mapped.ILLUMINA...\n",
       "3    data/HG00100/alignment/HG00100.mapped.ILLUMINA...\n",
       "4    data/HG00101/alignment/HG00101.mapped.ILLUMINA...\n",
       "Name: BAM FILE, dtype: object"
      ]
     },
     "execution_count": 81,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# filter index file for mapped reads only\n",
    "paths = pd.read_csv(idx_1kg_phase3,sep='\\t').iloc[:,0]\n",
    "mask = paths.str.contains(r'\\.mapped\\.ILLUMINA.*bam$',regex=True)\n",
    "print(mapped.shape)\n",
    "mapped.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 98,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "'s3://1000genomes/phase3/data/HG01125/alignment/HG01125.mapped.ILLUMINA.bwa.CLM.low_coverage.20120522.bam'"
      ]
     },
     "execution_count": 98,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "def download_1kg_bam_region(\n",
    "    index_series,\n",
    "    sample,\n",
    "    chrom,\n",
    "    left,\n",
    "    right,\n",
    "    outfile,\n",
    "    bucket='1000genomes',\n",
    "    phase='phase3'\n",
    "):\n",
    "    chrom = str(chrom)\n",
    "    left = int(left)\n",
    "    right = int(right)\n",
    "\n",
    "    # filter index for sample\n",
    "    mask = index_series.str.contains(sample)\n",
    "    if bool(mask.sum() > 1):\n",
    "        raise ValueError(\"sample name matched more than one file in index\")\n",
    "    key_bam = index_series[mask].tolist()[0]\n",
    "    key_bam = os.path.join(phase, key_bam) # e.g., phase3/data/HG01125/alignment/HG01125.mapped.ILLUMINA.bwa.CLM.low_coverage.20120522.bam\n",
    "    key_bai = key_bam + '.bai'\n",
    "    # e.g., s3://1000genomes/phase3/data/HG01125/alignment/HG01125.mapped.ILLUMINA.bwa.CLM.low_coverage.20120522.bam\n",
    "    url_bam = os.path.join(\n",
    "        's3://', bucket, key_bam\n",
    "    )\n",
    "    return url_bam\n",
    "\n",
    "    base = 'test'\n",
    "    out_idx=os.path.join(scratch, f'{base}.full.bam.bai')\n",
    "    #url_idx=\"s3://1000genomes/phase3/data/HG01125/alignment/HG01125.mapped.ILLUMINA.bwa.CLM.low_coverage.20120522.bam.bai\"\n",
    "    url_bam=\"s3://1000genomes/phase3/data/HG01125/alignment/HG01125.mapped.ILLUMINA.bwa.CLM.low_coverage.20120522.bam\"\n",
    "\n",
    "    # download index\n",
    "    s3 = boto3.client('s3')\n",
    "    s3.download_file(bucket, key_idx, out_idx)\n",
    "\n",
    "    with pysam.AlignmentFile(url_bam, \"rb\", index_filename=out_idx) as f_in:\n",
    "        with pysam.AlignmentFile(out, \"wb\", header=f_in.header) as f_out:\n",
    "            for read in f_in.fetch(chrom, region_l, region_r):\n",
    "                f_out.write(read)\n",
    "    # index region bam\n",
    "    pysam.index(out)\n",
    "download_1kg_bam_region(mapped,\"HG01125\",19,54886338,54889354,'test.py.bam')\n",
    "        "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 44,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>id</th>\n",
       "      <th>n_modes</th>\n",
       "      <th>chr</th>\n",
       "      <th>start</th>\n",
       "      <th>stop</th>\n",
       "      <th>allele_freq</th>\n",
       "      <th>length</th>\n",
       "      <th>n_samples</th>\n",
       "      <th>n_ref_samples</th>\n",
       "      <th>mode_1</th>\n",
       "      <th>mode_2</th>\n",
       "      <th>mode_3</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>UW_VH_9038</td>\n",
       "      <td>1</td>\n",
       "      <td>19</td>\n",
       "      <td>54887338</td>\n",
       "      <td>54888354</td>\n",
       "      <td>0.014</td>\n",
       "      <td>1016</td>\n",
       "      <td>51</td>\n",
       "      <td>0</td>\n",
       "      <td>['HG01125', 'HG01896', 'HG01956', 'HG02009', '...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>SI_BD_10797</td>\n",
       "      <td>1</td>\n",
       "      <td>10</td>\n",
       "      <td>86801825</td>\n",
       "      <td>86802449</td>\n",
       "      <td>0.033</td>\n",
       "      <td>624</td>\n",
       "      <td>283</td>\n",
       "      <td>139</td>\n",
       "      <td>['HG00154', 'HG00159', 'HG00250', 'HG00264', '...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>UW_VH_19141</td>\n",
       "      <td>1</td>\n",
       "      <td>3</td>\n",
       "      <td>177294474</td>\n",
       "      <td>177297489</td>\n",
       "      <td>0.032</td>\n",
       "      <td>3015</td>\n",
       "      <td>133</td>\n",
       "      <td>1</td>\n",
       "      <td>['HG00180', 'HG00185', 'HG00189', 'HG00237', '...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>DEL_pindel_47187</td>\n",
       "      <td>2</td>\n",
       "      <td>18</td>\n",
       "      <td>45379612</td>\n",
       "      <td>45379612</td>\n",
       "      <td>0.550</td>\n",
       "      <td>195</td>\n",
       "      <td>847</td>\n",
       "      <td>66</td>\n",
       "      <td>['HG00114', 'HG00132', 'HG00142', 'HG00150', '...</td>\n",
       "      <td>['HG00108', 'HG00111', 'HG00121', 'HG00125', '...</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>DEL_pindel_24042</td>\n",
       "      <td>2</td>\n",
       "      <td>7</td>\n",
       "      <td>136996507</td>\n",
       "      <td>136996739</td>\n",
       "      <td>0.071</td>\n",
       "      <td>232</td>\n",
       "      <td>184</td>\n",
       "      <td>48</td>\n",
       "      <td>['HG01241', 'HG01259', 'HG01488', 'HG01890', '...</td>\n",
       "      <td>['HG01125', 'HG01392', 'HG01403', 'HG01556', '...</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                 id  n_modes  chr      start       stop  allele_freq  length  \\\n",
       "0        UW_VH_9038        1   19   54887338   54888354        0.014    1016   \n",
       "1       SI_BD_10797        1   10   86801825   86802449        0.033     624   \n",
       "2       UW_VH_19141        1    3  177294474  177297489        0.032    3015   \n",
       "3  DEL_pindel_47187        2   18   45379612   45379612        0.550     195   \n",
       "4  DEL_pindel_24042        2    7  136996507  136996739        0.071     232   \n",
       "\n",
       "   n_samples  n_ref_samples  \\\n",
       "0         51              0   \n",
       "1        283            139   \n",
       "2        133              1   \n",
       "3        847             66   \n",
       "4        184             48   \n",
       "\n",
       "                                              mode_1  \\\n",
       "0  ['HG01125', 'HG01896', 'HG01956', 'HG02009', '...   \n",
       "1  ['HG00154', 'HG00159', 'HG00250', 'HG00264', '...   \n",
       "2  ['HG00180', 'HG00185', 'HG00189', 'HG00237', '...   \n",
       "3  ['HG00114', 'HG00132', 'HG00142', 'HG00150', '...   \n",
       "4  ['HG01241', 'HG01259', 'HG01488', 'HG01890', '...   \n",
       "\n",
       "                                              mode_2 mode_3  \n",
       "0                                                NaN    NaN  \n",
       "1                                                NaN    NaN  \n",
       "2                                                NaN    NaN  \n",
       "3  ['HG00108', 'HG00111', 'HG00121', 'HG00125', '...    NaN  \n",
       "4  ['HG01125', 'HG01392', 'HG01403', 'HG01556', '...    NaN  "
      ]
     },
     "execution_count": 44,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "2533    data/NA21143/alignment/NA21143.mapped.ILLUMINA...\n",
       "Name: BAM FILE, dtype: object"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "mapped.str.contains('NA21143')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 28,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "''"
      ]
     },
     "execution_count": 28,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "import boto3\n",
    "import pysam\n",
    "import os\n",
    "\n",
    "base = 'test'\n",
    "bucket='1000genomes'\n",
    "key_idx='phase3/data/HG01125/alignment/HG01125.mapped.ILLUMINA.bwa.CLM.low_coverage.20120522.bam.bai'\n",
    "out_idx=os.path.join(scratch, f'{base}.bam.bai')\n",
    "#url_idx=\"s3://1000genomes/phase3/data/HG01125/alignment/HG01125.mapped.ILLUMINA.bwa.CLM.low_coverage.20120522.bam.bai\"\n",
    "url_bam=\"s3://1000genomes/phase3/data/HG01125/alignment/HG01125.mapped.ILLUMINA.bwa.CLM.low_coverage.20120522.bam\"\n",
    "chrom = \"19\"\n",
    "region_l = 54886338\n",
    "region_r= 54889354\n",
    "out_bam=os.path.join(scratch, f'{base}.bam')\n",
    "\n",
    "# download index\n",
    "s3 = boto3.client('s3')\n",
    "s3.download_file(bucket, key_idx, out_idx)\n",
    "\n",
    "## Use pysam's remote access capability\n",
    "#url = f\"s3://{bucket_name}.s3.amazonaws.com/{bam_key}\"\n",
    "with pysam.AlignmentFile(url_bam, \"rb\", index_filename=out_idx) as f_in:\n",
    "    with pysam.AlignmentFile(out_bam, \"wb\", header=f_in.header) as f_out:\n",
    "        for read in f_in.fetch(chrom, region_l, region_r):\n",
    "            f_out.write(read)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 29,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "167.27333333333337"
      ]
     },
     "execution_count": 29,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "2.2 * 4562 / 60"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.12.2"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
